consoles/sshVirtsh.pm

```diff
@@ -325,111 +325,159 @@ sub add_interface {
+            sleep 5;
+        }
+        last unless $ret;
+    }
+    die "Too many attempts to format HDD" if $stderr;
+    bmwqemu::diag("gen_kvm_img: Finished creating image file...");
+}
+
+=head2 download_image_file
+
+Internal method to download asset files using rsync. Requires absolute path to source file
+and destination directory. If successful, returns absolute path to downloaded file
+
+    $self->download_image_file('/var/lib/openqa/share/factory/hdd/SUSE-MicroOS.s390x-5.1.0-Default-kvm-Build10.9.qcow2', '/var/lib/libvirt/images/');
+
+=cut
+
+sub download_image_file {
+    my ($self, $src, $dest) = @_;
+    $dest .= basename($src);
+    die "Downloading $src failed!" if $self->run_cmd("rsync -av $src $dest");
+    return $dest;
+}
+
+sub decompress_cmd {
+    return sprintf('xz --decompress --keep --verbose --no-warn --quiet %s', shift);
+}
+
 sub add_disk {
     my ($self, $args) = @_;
+    die('Use one out of `backingfile`, `create` or `cdrom`') if ((($args->{backingfile} // 0) + ($args->{cdrom} // 0) + ($args->{create} // 0)) != 1);
+    die('`backingfile` or `cdrom` has been used without a given image file') if (($args->{backingfile} || $args->{cdrom}) && !$args->{file});
 
-    my $backingfile             = $args->{backingfile};
-    my $cdrom                   = $args->{cdrom};
-    my $name                    = $self->name;
-    my $file                    = $name . $args->{dev_id} . ($self->vmm_family eq 'vmware' ? '.vmdk' : '.img');
-    my $basedir                 = '/var/lib/libvirt/images/';
-    my $vmware_datastore        = get_var('VMWARE_DATASTORE', '');
-    my $vmware_openqa_datastore = "/vmfs/volumes/$vmware_datastore/openQA/";
-    if ($args->{create}) {
-        my $size = $args->{size} || '20G';
-        if ($self->vmm_family eq 'vmware') {
-            my $vmware_disk_path = $vmware_openqa_datastore . $file;
-            # Power VM off, delete it's disk image, and create it again.
-            # Than wait for some time for the VM to *really* turn off.
-            my $cmd =
-              "( set -x; vmid=\$(vim-cmd vmsvc/getallvms | awk \'/$name/ { print \$1 }\');" .
-              'if [ $vmid ]; then ' .
-              'vim-cmd vmsvc/power.off $vmid;' .
-              'vim-cmd vmsvc/destroy $vmid;' .
-              'fi;' .
-              "vmkfstools -v1 -U $vmware_disk_path;" .
-              "vmkfstools -v1 -c $size --diskformat thin $vmware_disk_path; sleep 10 ) 2>&1";
-            my $retval = $self->run_cmd($cmd, domain => 'sshVMwareServer');
-            die "Can't create VMware image $vmware_disk_path" if $retval;
+    my $size = $args->{size} // '20G';
+    $size .= 'G' if ($size =~ /^\d+$/);
+
+    my $basedir;
+    my $file;
+    my $vmware_datastore = get_var('VMWARE_DATASTORE', '');
+
+    if ($self->vmm_family eq 'vmware') {
+        my $basedir = "/vmfs/volumes/$vmware_datastore/openQA/";
+        $file = sprintf("%s%s.vmdk", $self->{name}, $args->{dev_id});
```

Comment:
`$args->{dev_id}` can be `undef` at this point. Maybe we need some input validation for it as well?

Reply:
Seems like another mandatory argument that has to be passed :)
I am briefly checking the callers, seems like it is always there.

Reply:
and we would create a problem over here...

```
/lib/bootloader_setup.pm:            $svirt->add_disk({file => $patched_img, dev_id => 'a'});
```

Reply:
ah because we miss `create` `cdrom` or `backingfile`... what was the default before?

Reply:
> ah because we miss `create` `cdrom` or `backingfile`... what was the default before?

no, we haven't missed anything. That caller comes from test repo, it is actually the only one that I could have found that wants to use the original image :/

Reply:
As I understand this tree https://openqa.suse.de/tests/6901438#dependencies, they use single image without backing file.

Reply:
> As I understand this tree https://openqa.suse.de/tests/6901438#dependencies, they use single image without backing file.

I am wrong, `backend::baseclass::run_ssh_cmd(cmd="qemu-img create '/var/lib/libvirt/images/openQA-SUT-6a.img' -f qcow2 -b '/var/lib/libvirt/images//SLES-15-SP3-s390x-mru-install-minimal-with-addons-Build20210820-1-Server-DVD-Updates-s390x-kvm-sle12.qcow2' 42949672960"` Sorry for the noise

Reply:
I think we would break this part https://openqa.suse.de/tests/6892091

Reply:
I'm still not sure what we did wrong?

So my advice for the input validation was wrong, or?
And we need to copy the file if not `backing|create|cdrom` is given.

Reply:
We also miss a test-case for that

Reply:
> I'm still not sure what we did wrong?
> 
> So my advice for the input validation was wrong, or?
> And we need to copy the file if not `backing|create|cdrom` is given.

We have missed a valid migration scenario. They boot older sle, patch it, reboot so they initiate YaST2 installer to upgrade the system. 

So at the very beginning, they download image, create backing file for it. And at the second executeion of `bootloader_zkvm` they just use the XML part of the code

Reply:
who does the cleanup of the backing files?

Reply:
I am aware of only this one https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/installation/bootloader_svirt.pm#L69

Reply:
Not sure whether there is any clean up for xen or zkvm

Reply:
ok